### PR TITLE
[analyzer][NFC] Factor out NoOwnershipChangeVisitor

### PR DESCRIPTION
**Pull request by Szelethus**
In preparation for adding essentially the same visitor to StreamChecker, this patch factors this visitor out to a common header.

I'll be the first to admit that the interface of these classes are not terrific, but it rather tightly held back by its main technical debt, which is NoStoreFuncVisitor, the main descendant of NoStateChangeVisitor.

Change-Id: I99d73ccd93a18dd145bbbc83afadbb432dd42b90

### CODE REVIEW

**Comment by Szelethus** [comment]
Yes, it is, sorry about that :)

**Review by NagyDonat — APPROVED** [review]
Perfectly normal NFC change, feel free to merge it (assuming that the tests pass).

**Review by NagyDonat — APPROVED** [review]

File: clang/lib/StaticAnalyzer/Checkers/NoOwnershipChangeVisitor.cpp [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+#include "clang/StaticAnalyzer/Core/PathSensitive/ProgramState_Fwd.h"
+#include "llvm/ADT/SetOperations.h"
+
+using namespace clang;
+using namespace ento;
+using OwnerSet = NoOwnershipChangeVisitor::OwnerSet;
+
+// Collect which entities point to the allocated memory, and could be
+// responsible for deallocating it.
+class OwnershipBindingsHandler : public StoreManager::BindingsHandler {
```
Comment (balazske) [inline clang/lib/StaticAnalyzer/Checkers/NoOwnershipChangeVisitor.cpp:23]: This class can be in an anonymous namespace?

File: clang/lib/StaticAnalyzer/Checkers/MallocChecker.cpp [hunk 2]
```diff
@@ -804,23 +751,21 @@ class NoOwnershipChangeVisitor final : public NoStateChangeFuncVisitor {
...
+
   /// Heuristically guess whether the callee intended to free memory. This is
   /// done syntactically, because we are trying to argue about alternative
   /// paths of execution, and as a consequence we don't have path-sensitive
   /// information.
-  bool doesFnIntendToHandleOwnership(const Decl *Callee, ASTContext &ACtx) {
+  bool doesFnIntendToHandleOwnership(const Decl *Callee,
+                                     ASTContext &ACtx) final {
     using namespace clang::ast_matchers;
     const FunctionDecl *FD = dyn_cast<FunctionDecl>(Callee);
```
Comment (mikerice1969) [inline clang/lib/StaticAnalyzer/Checkers/MallocChecker.cpp:767]: Following up on static verifier hits. This dyn_cast result is now not checked before it is dereferenced below (FD->getBody()). Should this be a cast instead? Or should there be a check before the dereference?

Comment (Szelethus) [reply clang/lib/StaticAnalyzer/Checkers/MallocChecker.cpp:767]: Nice catch! Fix is here: https://github.com/llvm/llvm-project/pull/100719

**Pull request merged**